src/index.ts

```diff
@@ -184,6 +186,185 @@ export function HTPipeAttachData<
+                  ? TLeft['finalAuthorize']
+                  : () => {}
+              >
+            >
+          >
+        : any)
+  >,
+  TRight extends MightHaveFinalAuthorize<any, any>,
+  TContextInLeft extends Parameters<TLeft['finalAuthorize']>[0],
+  TContextOutLeft extends PromiseResolveOrSync<
+    ReturnType<TLeft['finalAuthorize']>
+  >
+>(
+  left: TLeft,
+  right: TRight
+): HasFinalAuthorize<TContextInLeft, TContextOutLeft>;
+
+// right has finalAuthorize and left does not
+export function HTPipeFinalAuthorize<
+  TLeft extends MightHaveFinalAuthorize<
+    any,
+    | boolean
+    | (TRight extends HasFinalAuthorize<any, any>
+        ? Pick<
+            Parameters<TRight['finalAuthorize']>[0],
+            keyof PromiseResolveOrSync<
+              ReturnType<
+                TLeft extends HasFinalAuthorize<any, any>
+                  ? TLeft['finalAuthorize']
+                  : () => {}
+              >
+            >
+          >
+        : any)
+  >,
+  TRight extends HasFinalAuthorize<any, any>,
+  TContextInRight extends Parameters<TRight['finalAuthorize']>[0],
+  TContextOutRight extends PromiseResolveOrSync<
+    ReturnType<TRight['finalAuthorize']>
+  >
+>(
+  left: TLeft,
+  right: TRight
+): HasFinalAuthorize<TContextInRight, TContextOutRight>;
+
+// right and left doesn't have preAuthorize
+export function HTPipeFinalAuthorize<
+  TLeft extends MightHaveFinalAuthorize<
+    any,
+    | boolean
+    | (TRight extends HasFinalAuthorize<any, any>
+        ? Pick<
+            Parameters<TRight['finalAuthorize']>[0],
+            keyof PromiseResolveOrSync<
+              ReturnType<
+                TLeft extends HasFinalAuthorize<any, any>
+                  ? TLeft['finalAuthorize']
+                  : () => {}
+              >
+            >
+          >
+        : any)
+  >,
+  TRight extends MightHaveFinalAuthorize<any, any>
+>(left: TLeft, right: TRight): {};
+
+// finalAuthorize main function
+export function HTPipeFinalAuthorize<
+  TLeft extends MightHaveFinalAuthorize<any, any>,
+  TRight extends MightHaveFinalAuthorize<any, any>,
+  TContextInLeft extends TLeft extends HasFinalAuthorize<any, any>
+    ? Parameters<TLeft['finalAuthorize']>[0]
+    : never,
+  TContextInRight extends TRight extends HasFinalAuthorize<any, any>
+    ? Parameters<TRight['finalAuthorize']>[0]
+    : never,
+  TContextOutLeft extends TLeft extends HasFinalAuthorize<any, any>
+    ? PromiseResolveOrSync<ReturnType<TLeft['finalAuthorize']>>
+    : never,
+  TContextOutRight extends TRight extends HasFinalAuthorize<any, any>
+    ? PromiseResolveOrSync<ReturnType<TRight['finalAuthorize']>>
+    : never
+>(left: TLeft, right: TRight) {
+  if (isHasFinalAuthorize(left) && isHasFinalAuthorize(right)) {
+    return {
+      finalAuthorize: async (
+        context: TContextOutLeft extends TContextInRight
+          ? TContextInLeft
+          : TContextInRight & TContextInLeft
+      ) => {
```

Comment:
this seems different than the other one - for example, the || false.  why?
- [ ] make it similar to preAuthorize

Reply:
done